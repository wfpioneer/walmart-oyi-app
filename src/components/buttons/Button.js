/* eslint react/sort-comp: 0 */
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import COLOR from '../../themes/Color';

/**
 * Custom button component
 *
 */

class Button extends PureComponent {
  static Type = {
    PRIMARY: 1,
    SOLID_WHITE: 2,
    NO_BORDER: 3
  };

  constructor(props) {
    super(props);
    const { backgroundColor } = this.props;
    this.bgColor = backgroundColor;
  }

  onPress = () => {
    const { onPress } = this.props;
    if (onPress !== undefined) {
      onPress();
    }
  };

  render() {
    const {
      type,
      width,
      height,
      disabled,
      backgroundColor,
      radius,
      style,
      titleColor,
      titleFontSize,
      titleFontWeight,
      title,
      testID
    } = this.props;
    const containerStyle = width === -1
      ? { height }
      : {
        height,
        width
      };

    this.bgColor = disabled ? COLOR.DISABLED_BLUE : backgroundColor;

    let buttonStyle;
    if (type === Button.Type.PRIMARY) {
      buttonStyle = {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: radius,
        backgroundColor: this.bgColor,
        elevation: 1
      };
    } else if (type === Button.Type.SOLID_WHITE) {
      buttonStyle = {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: radius,
        borderColor: this.bgColor,
        borderWidth: 1,
        backgroundColor: COLOR.WHITE
      };
    } else if (type === Button.Type.NO_BORDER) {
      buttonStyle = {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: COLOR.TRANSPARENT
      };
    }
    return (
      <View style={[containerStyle, style]}>
        <View style={{ flex: 1 }}>
          <TouchableOpacity
            disabled={disabled}
            onPress={this.onPress}
            activeOpacity={0.8}
            style={buttonStyle}
            testID={testID}
          >
            <Text
              style={{
                color: titleColor,
                fontSize: titleFontSize,
                fontFamily: 'Roboto-Medium',
                fontWeight: titleFontWeight
              }}
            >
              {title}
            </Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}

Button.propTypes = {
  onPress: PropTypes.func,
  title: PropTypes.string,
  titleColor: PropTypes.string,
  titleFontSize: PropTypes.number,
  titleFontWeight: PropTypes.string,
  backgroundColor: PropTypes.string,
  height: PropTypes.number,
  width: PropTypes.number,
  radius: PropTypes.number,
  style: PropTypes.object,
  type: PropTypes.number,
  disabled: PropTypes.bool,
  testID: PropTypes.string
};

Button.defaultProps = {
  onPress: undefined,
  title: 'default',
  titleColor: COLOR.WHITE,
  titleFontSize: 16,
  titleFontWeight: 'normal',
  backgroundColor: COLOR.MAIN_THEME_COLOR,
  height: 45,
  width: -1,
  radius: 5,
  style: undefined,
  type: 1,
  disabled: false
};

export default Button;
