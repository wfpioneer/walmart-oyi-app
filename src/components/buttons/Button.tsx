import React from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import COLOR from '../../themes/Color';

/**
 * Custom button component
 *
 */

// eslint-disable-next-line no-shadow
export enum ButtonType {
  PRIMARY = 1,
  SOLID_WHITE = 2,
  NO_BORDER = 3
}

type fontWeightType = 'normal' | 'bold' | '100' | '200' | '300' |
    '400' | '500' | '600' | '700' | '800' | '900' | undefined;

type textAlignType = 'auto' | 'left' | 'right' | 'center' | 'justify' | undefined;

interface ButtonPropsI {
  onPress?: () => void,
  title?: string,
  titleColor?: string,
  titleFontSize?: number,
  titleFontWeight?: fontWeightType,
  titleAlign?: textAlignType,
  backgroundColor?: string,
  height?: number | string,
  width?: number | string,
  radius?: number,
  style?: object,
  type?: number,
  disabled?: boolean,
  // eslint-disable-next-line react/require-default-props
  testID?: string
}

export const Button = (props: ButtonPropsI): JSX.Element => {
  const {
    type,
    width,
    height,
    disabled,
    backgroundColor,
    radius,
    style,
    titleColor,
    titleFontSize,
    titleFontWeight,
    titleAlign,
    title,
    testID
  } = props;

  const onButtonPress = () => {
    const { onPress } = props;
    if (onPress !== undefined) {
      onPress();
    }
  };

  const containerStyle = width === -1
    ? { height }
    : {
      height,
      width
    };
  const bgColor = disabled ? COLOR.DISABLED_BLUE : backgroundColor;

  let buttonStyle = {};
  if (type === ButtonType.PRIMARY) {
    buttonStyle = {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: radius,
      backgroundColor: bgColor,
      elevation: 1
    };
  } else if (type === ButtonType.SOLID_WHITE) {
    buttonStyle = {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: radius,
      borderColor: bgColor,
      borderWidth: 1,
      backgroundColor: COLOR.WHITE
    };
  } else if (type === ButtonType.NO_BORDER) {
    buttonStyle = {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: COLOR.TRANSPARENT
    };
  }
  return (
    <View style={[containerStyle, style]}>
      <View style={{ flex: 1 }}>
        <TouchableOpacity
          disabled={disabled}
          onPress={onButtonPress}
          activeOpacity={0.8}
          style={buttonStyle}
          testID={testID}
        >
          <Text
            style={{
              color: titleColor,
              fontSize: titleFontSize,
              fontFamily: 'Roboto-Medium',
              fontWeight: titleFontWeight,
              textAlign: titleAlign
            }}
          >
            {title}
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

Button.defaultProps = {
  onPress: undefined,
  title: 'default',
  titleColor: COLOR.WHITE,
  titleFontSize: 16,
  titleFontWeight: 'normal',
  titleAlign: undefined,
  backgroundColor: COLOR.MAIN_THEME_COLOR,
  height: 45,
  width: -1,
  radius: 5,
  style: undefined,
  type: 1,
  disabled: false
};

export default Button;
