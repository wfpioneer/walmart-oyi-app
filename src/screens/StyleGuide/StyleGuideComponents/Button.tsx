import React from 'react';
import { View } from 'react-native';
import Button from '../../../components/button/Button';
import COLOR from '../../../themes/Color';

export default class ButtonStyleGuide extends React.PureComponent {
  render(): React.ReactElement<any> {
    return (
      <View>
        <Button type={Button.Type.PRIMARY} title="Primary button type" />
        <Button type={Button.Type.NO_BORDER} title="No border button" titleColor={COLOR.MAIN_THEME_COLOR} />
        <Button type={Button.Type.SOLID_WHITE} title="Solid white button" titleColor={COLOR.MAIN_THEME_COLOR} />
        <Button type={Button.Type.PRIMARY} title="Disabled button" disabled />
      </View>
    );
  }
}
