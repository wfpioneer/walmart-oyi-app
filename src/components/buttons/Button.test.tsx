import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import COLOR from '../../themes/Color';
import { Button, ButtonType } from './Button';

describe('Button Component', () => {
  it('Renders the default button', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it('Renders the button with type as Primary', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button type={ButtonType.PRIMARY} />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it('Renders the button with type as Solid White', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button type={ButtonType.SOLID_WHITE} />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it('Renders the button with type as No Border', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button type={ButtonType.NO_BORDER} />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it('Renders the disabled button', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button disabled />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it('Renders the button with other custom specification props', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <Button
        onPress={jest.fn}
        title="Test Button"
        titleColor={COLOR.WHITE}
        titleFontSize={12}
        titleFontWeight="normal"
        backgroundColor={COLOR.MAIN_THEME_COLOR}
        height={5}
        width={10}
        radius={2}
        style={{}}
        testID="testButton"
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('Tests button click functionality', () => {
    const mockOnPressCallBackFn = jest.fn();
    const { getByTestId } = render(
      <Button
        testID="testButton"
        onPress={mockOnPressCallBackFn}
      />
    );
    const button = getByTestId('testButton');
    fireEvent.press(button);
    expect(mockOnPressCallBackFn).toBeCalledTimes(1);
  });

  it('Tests button title been correctly displayed', () => {
    const mockOnPressCallBackFn = jest.fn();
    const { queryAllByText } = render(
      <Button
        testID="testButton"
        title="Test Title"
        onPress={mockOnPressCallBackFn}
      />
    );
    expect(queryAllByText('Test Title')).toHaveLength(1);
  });
});
