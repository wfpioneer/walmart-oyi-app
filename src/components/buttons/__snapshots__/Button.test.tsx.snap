// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Renders the button with other custom specification props 1`] = `
<View
  style={
    Array [
      Object {
        "height": 5,
        "width": 10,
      },
      Object {},
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={false}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0071e9",
          "borderRadius": 2,
          "elevation": 1,
          "flex": 1,
          "justifyContent": "center",
        }
      }
      testID="testButton"
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 12,
            "fontWeight": "normal",
          }
        }
      >
        Test Button
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button Component Renders the button with type as No Border 1`] = `
<View
  style={
    Array [
      Object {
        "height": 45,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={false}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#00000000",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        default
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button Component Renders the button with type as Primary 1`] = `
<View
  style={
    Array [
      Object {
        "height": 45,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={false}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0071e9",
          "borderRadius": 5,
          "elevation": 1,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        default
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button Component Renders the button with type as Solid White 1`] = `
<View
  style={
    Array [
      Object {
        "height": 45,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={false}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffffff",
          "borderColor": "#0071e9",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        default
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button Component Renders the default button 1`] = `
<View
  style={
    Array [
      Object {
        "height": 45,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={false}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0071e9",
          "borderRadius": 5,
          "elevation": 1,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        default
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button Component Renders the disabled button 1`] = `
<View
  style={
    Array [
      Object {
        "height": 45,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.8}
      disabled={true}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#5FA7F2",
          "borderRadius": 5,
          "elevation": 1,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffffff",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        default
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
